pkg/spegel/spegel.go

```diff
@@ -0,0 +1,233 @@
+package spegel
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/k3s-io/k3s/pkg/clientaccess"
+	"github.com/k3s-io/k3s/pkg/daemons/config"
+	"github.com/k3s-io/k3s/pkg/version"
+	"github.com/rancher/dynamiclistener/cert"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/request/union"
+	"k8s.io/utils/pointer"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/stdr"
+	"github.com/gorilla/mux"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	ipfslog "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/p2p/host/peerstore/pstoreds"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/xenitab/spegel/pkg/oci"
+	"github.com/xenitab/spegel/pkg/registry"
+	"github.com/xenitab/spegel/pkg/routing"
+	"github.com/xenitab/spegel/pkg/state"
+)
+
+// DefaultRegistry is the default instance of a Spegel distributed registry
+var DefaultRegistry = &Config{
+	Bootstrapper: NewSelfBootstrapper(),
+	HandlerFunc: func(_ *Config, _ *mux.Router) error {
+		return errors.New("not implemented")
+	},
+	AuthFunc: func() authenticator.Request {
+		return union.New(nil)
+	},
+}
+
+var (
+	P2pAddressAnnotation = "p2p." + version.Program + ".cattle.io/node-address"
+	P2pEnabledLabel      = "p2p." + version.Program + ".cattle.io/enabled"
+)
+
+type authFunc func() authenticator.Request
+type handlerFunc func(config *Config, router *mux.Router) error
+
+// Config holds fields for a distributed registry
+type Config struct {
+	ClientCAFile   string
+	ClientCertFile string
+	ClientKeyFile  string
+
+	ServerCAFile   string
+	ServerCertFile string
+	ServerKeyFile  string
+
+	// ExternalAddress is the address for other nodes to connect to the registry API.
+	ExternalAddress string
+
+	// InternalAddress is the address for the local containerd instance to connect to the registry API.
+	InternalAddress string
+
+	// RegistryPort is the port for the registry API.
+	RegistryPort string
+
+	// PSK is the preshared key required to join the p2p network.
+	PSK []byte
+
+	// Bootstrapper is the bootstrapper that will be used to discover p2p peers.
+	Bootstrapper routing.Bootstrapper
+
+	// HandlerFunc will be called to add the registry API handler to an existing router.
+	HandlerFunc handlerFunc
+
+	// Authenticator will be called to retrieve an authenticator used to validate the request to the registry API.
+	AuthFunc authFunc
+}
+
+// These values are not currently configurable
+const (
+	resolveLatestTag  = false
+	resolveRetries    = 0
+	resolveTimeout    = time.Second * 5
+	registryScheme    = "https"
+	registryNamespace = "k8s.io"
+	routerPort        = "5001"
```

Comment:
Do you think that these values can be made configurable by using an override from say the `registries.yaml` in the future

```yaml
...
embedded-registry:
  router-port: 5001
```

Reply:
No, I don't think we should try to extend registries.yaml to contain field specific to spegel.

I don't really want to add a CLI flag for this either, but if I do it will be set server-side and passed down to clients, as I think it should be the same cluster-wide. Perhaps I would use an env var for it - `LIBP2P_PORT` or something.

Reply:
OK, I added an env var. I don't want to **ENCOURAGE** people to change it, but it is possible.

Reply:
Thanks Brandon. Introducing a new port can be hard as there are chances of conflict with another service. Making it configurable would be helpful in my case as we have moved from gravity to k3s and would like to reuse a port instead of asking customers to open another one.